tests/listusers/listusers.go

```diff
@@ -0,0 +1,199 @@
+package listusers
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"testing"
+
+	openfgav1 "github.com/openfga/api/proto/openfga/v1"
+	parser "github.com/openfga/language/pkg/go/transformer"
+	"github.com/stretchr/testify/require"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/status"
+	"sigs.k8s.io/yaml"
+
+	"github.com/openfga/openfga/assets"
+	listuserstest "github.com/openfga/openfga/internal/test/listusers"
+	"github.com/openfga/openfga/pkg/typesystem"
+
+	"github.com/openfga/openfga/pkg/tuple"
+	"github.com/openfga/openfga/tests/check"
+)
+
+var writeMaxChunkSize = 40 // chunk write requests into a chunks of this max size
+
+type individualTest struct {
+	Name   string
+	Stages []*stage
+}
+
+type listUsersTests struct {
+	Tests []individualTest
+}
+
+type stage struct {
+	Model               string
+	Tuples              []*openfgav1.TupleKey
+	ListUsersAssertions []*listuserstest.Assertion `json:"listUsersAssertions"`
+}
+
+type ClientInterface interface {
+	check.ClientInterface
+	ListUsers(ctx context.Context, in *openfgav1.ListUsersRequest, opts ...grpc.CallOption) (*openfgav1.ListUsersResponse, error)
+}
+
+// RunAllTests will invoke all list users tests.
+func RunAllTests(t *testing.T, client ClientInterface) {
+	t.Run("RunAll", func(t *testing.T) {
+		t.Run("ListUsers", func(t *testing.T) {
+			t.Parallel()
+			files := []string{
+				"tests/consolidated_1_1_tests.yaml",
+				"tests/abac_tests.yaml",
+			}
+
+			var allTestCases []individualTest
+
+			for _, file := range files {
+				var b []byte
+				var err error
+				b, err = assets.EmbedTests.ReadFile(file)
+				require.NoError(t, err)
+
+				var testCases listUsersTests
+				err = yaml.Unmarshal(b, &testCases)
+				require.NoError(t, err)
+
+				allTestCases = append(allTestCases, testCases.Tests...)
+			}
+
+			for _, test := range allTestCases {
+				test := test
+				runTest(t, test, client, false)
+				runTest(t, test, client, true)
+			}
+		})
+	})
+}
+
+func runTest(t *testing.T, test individualTest, client ClientInterface, contextTupleTest bool) {
+	ctx := context.Background()
+	name := test.Name
+
+	if contextTupleTest {
+		name += "_ctxTuples"
+	}
+
+	t.Run(name, func(t *testing.T) {
+		if contextTupleTest && len(test.Stages) > 1 {
+			// we don't want to run special contextual tuples test for these cases
+			// as multi-stages test has expectation tuples are in system
+			t.Skipf("multi-stages test has expectation tuples are in system")
+		}
+
+		t.Parallel()
+		resp, err := client.CreateStore(ctx, &openfgav1.CreateStoreRequest{Name: name})
+		require.NoError(t, err)
+
+		storeID := resp.GetId()
+
+		for stageNumber, stage := range test.Stages {
+			t.Run(fmt.Sprintf("stage_%d", stageNumber), func(t *testing.T) {
+				if contextTupleTest && len(stage.Tuples) > 20 {
+					// https://github.com/openfga/api/blob/05de9d8be3ee12fa4e796b92dbdd4bbbf87107f2/openfga/v1/openfga.proto#L151
```

Comment:
This is pointing to the wrong protobuf line.

It should be https://github.com/openfga/api/blob/591c8fa7765ca7b10feec619cf7d849380da3757/openfga/v1/openfga_service.proto#L877

Which highlights another problem - we're missing the validation constraints that require no more than 20 contextual tuples. We should have:

```protobuf
message ListUsersRequest {
    ...

    repeated TupleKey contextual_tuples =  6 [
        json_name = "contextual_tuples"
        (validate.rules).repeated.max_items = 20,
        (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_field) = {maximum: 20}
    ];
}
```

And that means we're also missing a test or two for that case.

Reply:
Addressed w/ https://github.com/openfga/api/pull/166. Also added test to this PR.